test/docs/jdk/javadoc/doccheck/doccheckutils/checkers/BadCharacterChecker.java

```diff
@@ -0,0 +1,158 @@
+    private static final Pattern metaCharset = Pattern.compile("(?i)<meta\\s+charset=\"([^\"]+)\">");
+    private static final Pattern metaContentType = Pattern.compile("(?i)<meta\\s+http-equiv=\"Content-Type\"\\s+content=\"text/html;charset=([^\"]+)\">");
+    private final Log errors;
+    private int files = 0;
+    private int badFiles = 0;
+
+    public BadCharacterChecker() {
+        errors = new Log();
+    }
+
+    public void checkFile(Path path) {
+        files++;
+        boolean ok = true;
+        try (InputStream in = new BufferedInputStream(Files.newInputStream(path))) {
+            CharsetDecoder d = getCharset(in).newDecoder()
+                    .onMalformedInput(CodingErrorAction.REPLACE)
+                    .onUnmappableCharacter(CodingErrorAction.REPLACE);
+            BufferedReader r = new BufferedReader(new InputStreamReader(in, d));
+            int lineNumber = 0;
+            String line;
+            try {
+                while ((line = r.readLine()) != null) {
+                    lineNumber++;
+                    int errorsOnLine = 0;
+                    for (int i = 0; i < line.length(); i++) {
+                        char ch = line.charAt(i);
+                        if (ch == 0xFFFD) {
+                            errorsOnLine++;
+                        }
+                    }
+                    if (errorsOnLine > 0) {
+                        errors.log(path, lineNumber, "found %d invalid characters", errorsOnLine);
+                        ok = false;
+                    }
+                }
+            } catch (IOException e) {
+                errors.log(path, lineNumber, e);
+                ok = false;
+
+            }
+        } catch (IOException e) {
+            errors.log(path, e);
+            ok = false;
+        }
+        if (!ok)
+            badFiles++;
+    }
+
+    @Override
+    public void checkFiles(List<Path> files) {
+        for (Path file : files) {
+            checkFile(file);
+        }
+    }
+
+    private Charset getCharset(InputStream in) throws IOException {
+        CharsetDecoder initial = StandardCharsets.US_ASCII.newDecoder()
+                .onMalformedInput(CodingErrorAction.REPLACE)
+                .onUnmappableCharacter(CodingErrorAction.REPLACE);
+
+        in.mark(1024);
+        try {
+            BufferedReader r = new BufferedReader(new InputStreamReader(in, initial));
+            char[] buf = new char[1024];
+            int n = r.read(buf, 0, buf.length);
+            String head = new String(buf, 0, n);
+            boolean html5 = doctype.matcher(head).find();
+            Matcher m1 = metaCharset.matcher(head);
+            if (m1.find()) {
+                return Charset.forName(m1.group(1));
+            }
+            Matcher m2 = metaContentType.matcher(head);
+            if (m2.find()) {
+                return Charset.forName(m2.group(1));
+            }
+            return html5 ? StandardCharsets.UTF_8 : StandardCharsets.ISO_8859_1;
```

Comment:
I assumed text would be written in latin characters, but I guess this can be removed and we can simply use UTF8?

Reply:
Unicode has some characters such as bidi characters which I don't want to allow but this test should only check for bistrot and character encoding, so UTF_8 could work as a default.

Reply:
Note that JDK source code is inconsistent.   The makefile commands for `javac` specify just ASCII (since forever) but the `javadoc` commands allow ISO-8859-1, which was popular before the widespread acceptance of UTF-8.

Ideally, all should be self-consistent -- JDK policy, `javac` commands, `javadoc` commands, DocCheck, etc.
Note that resource files might use different encodings to `*.java` source files.